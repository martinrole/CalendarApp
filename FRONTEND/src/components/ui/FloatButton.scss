.fab {
    border-radius: 100%;
    -webkit-border-radius: 100%;
    -moz-border-radius: 100%;
    -ms-border-radius: 100%;
    -o-border-radius: 100%;

    box-shadow: -1px 0px 10px 1px rgba(0, 0, 0, 0.7);
    -webkit-box-shadow: -1px 0px 10px 1px rgba(0, 0, 0, 0.7);
    -moz-box-shadow: -1px 0px 10px 1px rgba(0, 0, 0, 0.7);

    bottom: 25px;
    font-size: 25px;
    padding: 20px;
    position: fixed;
    right: 25px;
}