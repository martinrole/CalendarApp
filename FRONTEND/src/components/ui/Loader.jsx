import React from 'react'

export const Loader = () => {

    return (
        <div className="d-flex justify-content-center align-items-center" style={{ height: '100vh' }}>
            <div className="spinner-grow text-primary ml-3" role="status" />
            <div className="spinner-grow text-secondary ml-3" role="status" />
            <div className="spinner-grow text-success ml-3" role="status" />
            <div className="spinner-grow text-danger ml-3" role="status" />
            <div className="spinner-grow text-warning ml-3" role="status" />
            <div className="spinner-grow text-info ml-3" role="status" />
            <div className="spinner-grow text-light ml-3" role="status" />
            <div className="spinner-grow text-dark ml-3" role="status" />
            <span className="sr-only">Loading...</span>
        </div>
    )
}
