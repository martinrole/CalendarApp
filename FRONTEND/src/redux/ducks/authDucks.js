import { authFetch, authFetchToken } from "../../helpers/fetch";
import { showMessage } from "../../helpers/messages";


//1- initialState:
const initialState = {
    checking: true,
}

//2 - Types:
export const types = {
    checkingFinish: '[authDucks] Checking Finish',
    startLogin: '[authDucks] Start Login',
    login: '[authDucks] Login',
    startRegister: '[authDucks] Start Register',
    startTokenRenew: '[authDucks] Start Token renew',
    logout: '[authDucks] Logout'
}


//3 - Reducer:
export const authReducer = (state = initialState, action ) => {
    switch (action.type) {
        case types.login:
            return { 
                ...state, 
                ...action.payload,
                checking: false
            }
        case types.checkingFinish:
            return {
                ...state,
                checking: false
            }
        case types.logout:
            return {
                checking: false
            }
    
        default:
            return state;
    }
}

//4 - Actions:

//Primera forma de hacer:
// export const loginAction = ( correo, contrasena ) => {
//     return async ( dispatch ) => {
//         console.log('haber', correo, contrasena);
//     }
// }

//Segunda forma de hacerlo:
export const loginAction = ( correo, contrasena ) => async ( dispath, getState ) => {
    const data = { correo, contrasena }
    const { uid, nombre, token, ok, msg } = await authFetch('auth', data, 'POST')
    
    if ( ok ) {
        localStorage.setItem('token',token)
        localStorage.setItem('token-init-date', new Date().getTime())

        dispath({
            type: types.login,
            payload: {
                uid,
                nombre
            }
        })

    } else {
        showMessage('error', null, msg )
    }

}

export const registerAction = ( correo, name, contrasena ) => async ( dispath, getState) => {

    const data = { correo, nombre: name, contrasena }
    const { uid, nombre, token, ok, msg } = await authFetch('auth/create', data, 'POST') 

    if ( ok ) {
        localStorage.setItem('token',token)
        localStorage.setItem('token-init-date', new Date().getTime())

        dispath({
            type: types.login,
            payload: {
                uid,
                nombre
            }
        })

        showMessage('success', 'Bienvenido','Registro exitoso')

    } else {
        showMessage('error', null, msg )
    }
}

export const authCheckAction = () => async ( dispath, getState) => {

    const res = await authFetchToken('auth/updateToken', null , 'GET')
    
    if ( res.ok ) {
        localStorage.setItem('token',res.token)
        localStorage.setItem('token-init-date', new Date().getTime())

        dispath({
            type: types.login,
            payload: {
                uid: res.uid,
                nombre: res.nombre
            }
        })

    } else {
        dispath({
            type: types.checkingFinish
        })
    }
}

export const logoutAction = () =>  (dispath, getState) => {
    dispath({
        type: types.logout
    })
    
    localStorage.clear()
}