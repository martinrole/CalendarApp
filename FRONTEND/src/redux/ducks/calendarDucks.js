import { adjustEventsDate } from "../../helpers/calendar-config"
import { authFetchToken } from "../../helpers/fetch"
import { showMessage } from "../../helpers/messages"




// Lo use para antes de conectar con la base de datos:
// const initialState = {
//     events: [
//         {
//             id: 'sdsd&hnsd56gGRf123bSD',
//             title: 'PruebaReducer',
//             start: moment().toDate(),     //Sinónimo de new Date() de JavaScript
//             end: moment().add(1, 'hours').toDate(),
//             allDay: false,
//             notes: 'acá van las notas',
//             user: {
//                 _id: '12dsfv324D',
//                 nombre: 'Martin User'
//             }
//         }
//     ],
//     activeEvent: null
// }

//1- initialState:
const initialState = {
    events: [],
    activeEvent: null
}

//2 - Types:
const types = {
    eventNew: '[CalendarDucks] Add Event',
    eventList: '[CalendarDucks] List Event',
    eventActive: '[CalendarDucks] Active Event',
    eventClearActive: '[CalendarDucks] Clear Active Event',
    eventUpdate: '[CalendarDucks] Updated Event',
    eventDelete: '[CalendarDucks] Delete Event',
    clearCalendar: '[CalendarDucks] Clear All Calendar'
}

//3 - Reducer:
export const calendarReducer = (state = initialState, action ) => {
    switch (action.type) {
        case types.eventActive:
            return { 
                ...state, 
                activeEvent: action.payload 
            }
        case types.eventClearActive:
            return { 
                ...state, 
                activeEvent: null 
            }
        case types.eventNew:
            return { 
                ...state, 
                events: [...state.events, action.payload]
            }
        case types.eventList:
            return {
                ...state, 
                events: [ ...action.payload ] 
            }
        case types.eventUpdate:
            return {
                ...state, 
                events: state.events.map( event => event.id === action.payload.id ? action.payload : event )
            }
        case types.eventDelete:
            return {
                ...state, 
                events: state.events.filter( event => event.id !== state.activeEvent.id ),
                activeEvent: null
            }
        case types.clearCalendar:
            return {
                ...initialState
            }
        default:
            return state;
    }
}

//4 - Actions:
export const eventNewAction = ( event ) => async ( dispatch, getState ) => {
    
    const { uid, nombre } = getState().auth

    try {   
        const res = await authFetchToken('events/create', event, 'POST')

        const evento = {
            id: res.evento.id,
            ...event,
            usuario: {
                _id: uid,
                nombre
            }
        }
    
        if ( res.ok ) {
            dispatch({
                type: types.eventNew,
                payload: evento
            })
        }

    } catch (error) {
        console.log('error en eventNewAction', error);
        
    }
}


export const eventLoadAction = () => async ( dispatch, getState ) => {

    try {
        const { ok, eventos } = await authFetchToken('events/', null, 'GET')
        
        if (ok) {        
            const events = adjustEventsDate( eventos )
            dispatch({
                type: types.eventList,
                payload: events
            })
        }

    } catch (error) {
        console.log('error en eventLoadAction', error);
        
    }
}

export const eventUpdateAction = ( event ) => async (dispatch, getState ) => {

    try {
        const { ok, msg } = await authFetchToken(`events/${event.id}`, event, 'PUT')

        if ( ok ) {
            dispatch({
                type: types.eventUpdate,
                payload: event
            })
        } else {
            //Solo se pueden editar los eventos propios del user
            showMessage('error', null, msg)
        }

    } catch (error) {
        console.log('eventUpdateAction', error);
        
    }
}

export const eventDeleteAction = () => async (dispatch, getState ) => {

    try {
        const { id } = getState().calendar.activeEvent
        const { ok, msg } = await authFetchToken(`events/${id}`, {}, 'DELETE')

        if ( ok ) {
            dispatch({
                type: types.eventDelete
            })
        } else {
            showMessage('error', null, msg)
        }

    } catch (error) {
        console.log('error eventDeleteAction', error);   
    }
    
}

export const clearCalendarAction = () => (dispatch, getState ) => {
    console.log('clearCalendarAction: ', clearCalendarAction);
    dispatch({
        type: types.clearCalendar
    })
}

export const eventActiveAction = ( event ) => ({
    type: types.eventActive,
    payload: event
})

export const eventClearActiveAction = () => ({
    type: types.eventClearActive
})
