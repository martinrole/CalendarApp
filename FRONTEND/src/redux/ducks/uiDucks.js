

//1- initialState:
export const initialState = {
    modalOpen: false
}

//2 - Types:
export const types = {
    uiOpenModal: '[uiDucks] Open Modal',
    uiCloseModal: '[uiDucks] Close Modal'
}


//3 - Reducer:
export const uiReducer = (state = initialState, action ) => {
    switch (action.type) {
        case types.uiOpenModal:
            return { ...state, modalOpen: true }
        case types.uiCloseModal:
                return { ...state, modalOpen: false }
    
        default:
            return state;
    }
}

//4 - Actions:
export const openModalAction = () => ({
    type: types.uiOpenModal
})

export const closeModalAction = () => ({
    type: types.uiCloseModal
})