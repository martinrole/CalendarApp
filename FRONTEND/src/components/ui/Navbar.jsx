import { useDispatch, useSelector } from "react-redux";
import { useNavigate } from 'react-router-dom'
import { logoutAction } from "../../redux/ducks/authDucks";
import { clearCalendarAction } from "../../redux/ducks/calendarDucks";


export const Navbar = () => {

    const { nombre } = useSelector(state => state.auth)
    const dispatch = useDispatch()
    const navigate = useNavigate()

    const handleLogout = () => {
        navigate('/login', {
            replace: true
        })

        dispatch(logoutAction())
        dispatch(clearCalendarAction())
    }

    return (
        <div className="navbar navbar-dark bg-dark mb-4 px-3">
            <span className="navbar-brand">{nombre}</span>

            <button className="btn btn-outline-danger" onClick={handleLogout}>
                <span>Salir </span>
                <i className="fas fa-sign-out-alt"></i>
            </button>
        </div>
    )
}
