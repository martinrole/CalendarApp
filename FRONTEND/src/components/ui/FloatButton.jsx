import { useSelector } from "react-redux"
import { useDispatch } from "react-redux"
import { eventDeleteAction } from "../../redux/ducks/calendarDucks"
import { openModalAction } from "../../redux/ducks/uiDucks"


export const FloatButton = () => {

    const { activeEvent } = useSelector(state => state.calendar)

    const dispatch = useDispatch()

    const createEvent = () => {
        dispatch(openModalAction())
    }

    const deleteEvent = () => {
        dispatch(eventDeleteAction())
    }

    if (activeEvent === null) {
        //Crear evento
        return (
            <button className="btn btn-primary fab" onClick={createEvent}>
                <i className="fas fa-plus"></i>
            </button>
        )
    } else {
        //Eliminar evento
        return (
            <button className="btn btn-danger fab" onClick={deleteEvent}>
                <i className="fas fa-trash"></i>
            </button>
        )
    }
}
