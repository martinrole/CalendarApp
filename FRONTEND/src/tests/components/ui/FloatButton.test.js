import React from 'react'
import { mount } from 'enzyme'
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store' //ES6 modules
import thunk from 'redux-thunk';
import '@testing-library/jest-dom' // Solo sirve para tener el prototipado en vsCode
import { FloatButton } from '../../../components/ui/FloatButton';
import { eventDeleteAction } from '../../../redux/ducks/calendarDucks';
import { openModalAction } from '../../../redux/ducks/uiDucks';

const middlewares = [thunk]
const mockStore = configureStore(middlewares)


jest.mock('../../../redux/ducks/calendarDucks', () => ({
    eventDeleteAction: jest.fn()
}))

jest.mock('../../../redux/ducks/uiDucks', () => ({
    openModalAction: jest.fn()
}))


describe('Pruebas en componente <FloatButton />', () => {
    
    test('Debe mostrar FloatButton de eliminar evento', () => {

        const initialState = {
            calendar: {
                events: [],
                activeEvent: {
                    title: 'tituloNota',
                    start: 12433543,
                    end: 32544565,
                    id: 'idNota123'
                }
            }
        }

        const store = mockStore(initialState)

        const wrapper = mount(
            <Provider store={store}>
                <FloatButton />
            </Provider>
        )

        expect( wrapper ).toMatchSnapshot()
    });

    test('Debe llamar eventDeleteAction', () => {

        const initialState = {
            calendar: {
                events: [],
                activeEvent: {
                    title: 'tituloNota',
                    start: 12433543,
                    end: 32544565,
                    id: 'idNota123'
                }
            }
        }

        const store = mockStore(initialState)
        store.dispatch = jest.fn()

        const wrapper = mount(
            <Provider store={store}>
                <FloatButton />
            </Provider>
        )

        wrapper.find('button').simulate('click')
        expect( eventDeleteAction ).toHaveBeenCalledTimes(1)
    });

    test('Debe mostrar FloatButton de agregar evento', async () => {
        
        const initialState = {
            calendar: {
                events: [],
                activeEvent: null
            }
        }

        const store = mockStore(initialState)

        const wrapper = mount(
            <Provider store={store}>
                <FloatButton />
            </Provider>
        )

        expect( wrapper ).toMatchSnapshot()

    })

    test('Debe llamar createEvent y openModalAction', async () => {
        
        const initialState = {
            calendar: {
                events: [],
                activeEvent: null
            }
        }

        const store = mockStore(initialState)
        store.dispatch = jest.fn()

        const wrapper = mount(
            <Provider store={store}>
                <FloatButton />
            </Provider>
        )

        wrapper.find('button').simulate('click')
        expect( openModalAction ).toHaveBeenCalledTimes(1)
    })

})
