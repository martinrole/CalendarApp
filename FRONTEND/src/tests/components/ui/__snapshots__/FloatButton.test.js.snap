// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en componente <FloatButton /> Debe mostrar FloatButton de agregar evento 1`] = `
<button
  className="btn btn-primary fab"
  onClick={[Function]}
>
  <i
    className="fas fa-plus"
  />
</button>
`;

exports[`Pruebas en componente <FloatButton /> Debe mostrar FloatButton de eliminar evento 1`] = `
<button
  className="btn btn-danger fab"
  onClick={[Function]}
>
  <i
    className="fas fa-trash"
  />
</button>
`;
